### PR TITLE
Us397 develop

### PR DESCRIPTION
**Pull request by gangooteli**
This feature implements the minimum functionality of Result Generation 

- Height for a scan measure

Check the link for [dev build] (https://dev.azure.com/cgmorg/ChildGrowthMonitor/_releaseProgress?_a=release-pipeline-progress&releaseId=55)

- Able to generate the height prediction for the measure which are not in Dev environment
- Able to update the results in measure_result and artifact_Result table
- Able to create result in json file

### CODE REVIEW

File: src/result_gen.py [hunk 1]
```diff
@@ -0,0 +1,563 @@
...
+        print("Please provide model_id and endpoint name.")
+        exit(1)
+
+    # destination_folder = str(sys.argv[1])
+    # db_connection_file = str(sys.argv[2])
+    # storage_account_name = str(sys.argv[3])
+    model_id = str(sys.argv[1])
+    service = str(sys.argv[2])
+    # calibration_file = str(sys.argv[6])
+    # container_name = str(sys.argv[7])
```
Comment (markus-hinsche) [inline src/result_gen.py:478]: I am not a big fan of using `sys.argv` directly. 

I see the usage is:
`/usr/bin/python3 ~/src/result_gen.py 20190723-1119_2550-638weight aci-pointnet-weight-20190723-ci`

I would be more readible if we name the parameters, e.g.:
`/usr/bin/python3 ~/src/result_gen.py --model-id 20190723-1119_2550-638weight --service aci-pointnet-weight-20190723-ci`

File: src/result_gen.py [hunk 2]
```diff
@@ -0,0 +1,563 @@
...
+             ),
+            ("c66050300c1ab684_measure_1601355986152_WspeDENLEBteZLRD",
+             ),
+            ("c66050300c1ab684_measure_1602060079742_LORX5MwfKQhni3gA",
+             ),
+            ("601db192d38c0816_measure_1601379417732_PRBsdWChgw8Qkoe1",
+             )]
+        for id in measure_id:
+            id_split = id[0].split('_')
+            delete_measure_result = "delete from measure_result where measure_id = '{}'".format(
```
Comment (markus-hinsche) [inline src/result_gen.py:517]: `delete_measure_result` sounds like a name for a function

```suggestion
            sql_stmt_deletion = "delete from measure_result where measure_id = '{}'".format(
```

File: src/result_gen.py [hunk 3]
```diff
@@ -0,0 +1,563 @@
...
+                id[0]) + " and model_id = '{}';".format(model_id)
+            try:
+                main_connector.execute(delete_measure_result)
+            except Exception as error:
+                print(error)
+            delete_artifact_result = "delete from artifact_result where model_id = '{}'".format(
+                model_id) + " and artifact_id like '{}';".format(id_split[0] + "%" + id_split[2][:-1] + "%")
```
Comment (markus-hinsche) [inline src/result_gen.py:524]: Your string magic is a bit hard to read

Suggestion:
- Python3 has simplified this a lot, so we can use format strings

```suggestion
            tmp_str = id_split[0] + "%" + id_split[2][:-1] + "%"
            delete_artifact_result = f"delete from artifact_result where model_id = '{model_id}' and artifact_id like '{tmp_str}';"
```

File: src/result_gen.py [hunk 4]
```diff
@@ -0,0 +1,563 @@
...
+    Check the artifact belong to one of the status code in status list
+    '''
+
+    if extract_status_code(path) in status_list:
+        return True
+    else:
+        return False
+
+
+class Measure_Result_Generation:
```
Comment (markus-hinsche) [inline src/result_gen.py:85]:
```suggestion
class MeasureResultGeneration:
```

File: src/result_gen.py [hunk 5]
```diff
@@ -0,0 +1,563 @@
...
+
+def check_status_code(path, extract_status_code, status_list):
+    '''
+    Check the artifact belong to one of the status code in status list
+    '''
+
+    if extract_status_code(path) in status_list:
+        return True
+    else:
+        return False
```
Comment (markus-hinsche) [inline src/result_gen.py:82]: Personally I prefer to shorten code if it is still as readable afterwards

```suggestion
    return extract_status_code(path) in status_list:
```

File: src/result_gen.py [hunk 6]
```diff
@@ -0,0 +1,563 @@
...
+        self.ACC_NAME = config.ACC_NAME
+        self.ACC_KEY = config.ACC_KEY
+
+    def get_artifact_list_per_measure(self):
+        '''
+        Get the list of artifact per measure
+        '''
+        get_artifacts = "SELECT id, qr_code, create_timestamp, replace('{}'".format(
+            self.replace_path)
+        get_artifacts += " || split_part(storage_path,'/storage/emulated/0/Child Growth Monitor Scanner App/', 2), 'measurements', 'measure')"
```
Comment (markus-hinsche) [inline src/result_gen.py:116]: What is `/storage/emulated/0/Child Growth Monitor Scanner App/`. 
Consider putting it into a global variable

File: src/result_gen.py [hunk 7]
```diff
@@ -0,0 +1,563 @@
+import os
```
Comment (markus-hinsche) [inline src/result_gen.py:1]: This module is very large. Please consider splitting this up into submodules. 
Could we structure this along the different results? e.g.
- `height_results.py`
- `blur_results.py`

File: src/result_gen.py [hunk 8]
```diff
@@ -0,0 +1,563 @@
...
+import sys
+import json
+import shutil
+import random
+import numpy as np
+from bunch import Bunch
```
Comment (markus-hinsche) [inline src/result_gen.py:7]: I like bunch. 
Maybe we can use it instead `dotdict` https://github.com/Welthungerhilfe/cgm-ml/blob/9f031bbf2f8396fd568e9adcb0c3eb22a7b91a11/src/common/eval/QA/eval-depthmap-height/src/qa_config.py#L1

File: src/result_gen.py [hunk 9]
```diff
@@ -0,0 +1,563 @@
...
+        else:
+            self.dataformat = 'pcd'
+
+        self.replace_path = replace_path
+        self.container_name = container_name
+        self.front_status_list = ['100', '104', '200']
+        self.back_status_list = ['100', '104', '200']
+        self.threesixty_status_list = ['100', '104', '200']
+        self.ACC_NAME = config.ACC_NAME
+        self.ACC_KEY = config.ACC_KEY
```
Comment (markus-hinsche) [inline src/result_gen.py:108]: Instance variables should be lowercase

File: src/result_gen.py [hunk 10]
```diff
@@ -0,0 +1,563 @@
...
+                    self.artifact_back_pcd_paths, self.calibration, 'pointnet', input_shape)
+                threesixty_pcd_numpy = preprocessing.depthmap_to_pcd(
+                    self.artifact_threesixty_pcd_paths, self.calibration, 'pointnet', input_shape)
+            else:
+                front_pcd_numpy = preprocessing.pcd_to_ndarray(
+                    self.artifact_front_pcd_paths, input_shape)
+                back_pcd_numpy = preprocessing.pcd_to_ndarray(
+                    self.artifact_back_pcd_paths, input_shape)
+                threesixty_pcd_numpy = preprocessing.pcd_to_ndarray(
+                    self.artifact_threesixty_pcd_paths, input_shape)
```
Comment (markus-hinsche) [inline src/result_gen.py:306]: This code block looks very repetitive

File: src/result_gen.py [hunk 11]
```diff
@@ -0,0 +1,563 @@
...
+        self.back_predictions = inference.get_predictions_2(
+            self.artifact_back_pcd_numpy, model_id, service)
+        self.threesixty_predictions = inference.get_predictions_2(
+            self.artifact_threesixty_pcd_numpy, model_id, service)
+
+    def get_weight_per_artifact(self):
+        '''
+        Get the Weight prediction per artifact
+        '''
+        pass
```
Comment (markus-hinsche) [inline src/result_gen.py:327]: Should we remove this as it is unused. Maybe check for more occurrences of unused code

**Review by markus-hinsche — CHANGES_REQUESTED** [review]
I am not 100% sure how to check (try out) that this is working. 
Should we add unit tests for functions in `result_gen.py`?

File: src/result_gen.py (hunk 10 above)
Comment (gangooteli) [reply src/result_gen.py:302]: TBD later in next PR

File: src/result_gen.py (hunk 11 above)
Comment (gangooteli) [reply src/result_gen.py:323]: TBD later in next PR

File: src/result_gen.py (hunk 9 above)
Comment (gangooteli) [reply src/result_gen.py]: TBD later in next PR

File: src/result_gen.py (hunk 7 above)
Comment (gangooteli) [reply src/result_gen.py:1]: TBD later in next PR

**Comment by gangooteli** [comment]
For successful run of RG in dev environment, one can see this run in dev.
https://dev.azure.com/cgmorg/ChildGrowthMonitor/_releaseProgress?_a=release-pipeline-progress&releaseId=55.
Regd unit test, since there are lots of function used. We have started to write the unit test and will merge the code in next PR.

**Pull request merged**